### PR TITLE
Slot props reference block level variable

### PR DESCRIPTION
**Pull request by jasonlyu123**
#263 
Still to do
- [x] support component let
- [x] catch block
- [x] more test about variable scope 
- [x] more test about if identifier is member access

### CODE REVIEW

File: packages/svelte2tsx/src/nodes/TemplateScope.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    getInit(name: string): Identifier {
+        return this.inits.get(name) || this.parent?.getInit(name);
+    }
+
+	isLet(name: string) {
+		const owner = this.getOwner(name);
+		return owner && (owner.type === 'Element' || owner.type === 'InlineComponent');
+	}
+}
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/TemplateScope.ts:48]: Indentation is not consistent

File: packages/svelte2tsx/src/nodes/slot.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+export class SlotHandler {
+
+    constructor(private readonly str: MagicString, private readonly htmlx: string) { }
+
+    slots = new Map<string, Map<string, string>>();
+    resloved = new Map<Identifier, string>();
+    reslovedExpression = new Map<Node, string>();
+
+    reslove(identifierDef: Identifier, initExpression: Node, scope: TemplateScope) {
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/slot.ts:32]: `reslove` -> `resolve`. Happened in some other places as well 😄

File: packages/svelte2tsx/src/nodes/TemplateScope.ts (hunk 1 above)
Comment (jasonlyu123) [reply packages/svelte2tsx/src/nodes/TemplateScope.ts:48]: thanks!

File: packages/svelte2tsx/src/nodes/slot.ts (hunk 2 above)
Comment (jasonlyu123) [reply packages/svelte2tsx/src/nodes/slot.ts:32]: no wonder I keep not getting autocomplete 🤣

File: packages/svelte2tsx/src/nodes/slot.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
+import { Node, walk } from 'estree-walker';
+import MagicString from 'magic-string';
+import { attributeValueIsString, isMember } from '../utils/svelteAst';
+import TemplateScope from './TemplateScope';
+import { Identifier } from '../interfaces';
+
+function AttributeStrValueAsJsExpression(attr: Node): string {
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/slot.ts:7]: I prefer functions to be lowercased.

**Comment by dummdidumm** [comment]
Damn, that's fantastic work!

File: packages/svelte2tsx/src/nodes/TemplateScope.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        return this;
+    }
+
+    child() {
+        const child = new TemplateScope(this);
+        return child;
+    }
+
+    getOwner(name: string): Node {
+        return this.owners.get(name) || (this.parent && this.parent.getOwner(name));
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/TemplateScope.ts:37]: You can simplify this to

```suggestion
        return this.owners.get(name) || this.parent?.getOwner(name);
```

File: packages/svelte2tsx/src/nodes/slot.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    private getResolveExpressionStr(
+        identifierDef: Identifier,
+        scope: TemplateScope,
+        initExpression: Node
+    ) {
+        const { name } = identifierDef;
+
+        const owner = scope.getOwner(name);
```
Comment (dummdidumm) [inline packages/svelte2tsx/src/nodes/slot.ts:59]: `owner` is possibly undefined, another check is needed for this.

(note to self: we need to turn on `strictMode` someday for `svelte2tsx`)

**Review by dummdidumm — APPROVED** [review]
Awesome 👍

**Pull request merged**